File: src-tauri/src/commands/buffer.rs
use crate::error::AppError;
use crate::models::buffer::{
    CleanupTerminalBuffersRequest, GetTerminalBufferChunkRequest, GetTerminalBufferRequest,
    HasTerminalBufferRequest, TerminalBufferChunk,
};
use crate::services::buffer_manager::BufferStats;
use crate::state::AppState;
use tauri::State;

/// Get buffer as string for a terminal
#[tauri::command]
pub async fn get_terminal_buffer(
    request: GetTerminalBufferRequest,
    app_state: State<'_, AppState>,
) -> Result<String, AppError> {
    let buffer_manager = app_state.terminal_manager.get_buffer_manager();
    let buffer_string = buffer_manager.get_buffer_string(&request.terminal_id).await;
    Ok(buffer_string.unwrap_or_default())
}

/// Get buffer chunk for a terminal
#[tauri::command]
pub async fn get_terminal_buffer_chunk(
    request: GetTerminalBufferChunkRequest,
    app_state: State<'_, AppState>,
) -> Result<TerminalBufferChunk, AppError> {
    let buffer_manager = app_state.terminal_manager.get_buffer_manager();
    let chunk = buffer_manager
        .get_buffer_chunk(&request.terminal_id, request.start_line, request.chunk_size)
        .await;

    Ok(chunk)
}

/// Check if terminal has buffer
#[tauri::command]
pub async fn has_terminal_buffer(
    request: HasTerminalBufferRequest,
    app_state: State<'_, AppState>,
) -> Result<bool, AppError> {
    let buffer_manager = app_state.terminal_manager.get_buffer_manager();
    let has_buffer = buffer_manager.has_buffer(&request.terminal_id).await;
    Ok(has_buffer)
}

/// Get buffer statistics
#[tauri::command]
pub async fn get_buffer_stats(app_state: State<'_, AppState>) -> Result<BufferStats, AppError> {
    let buffer_manager = app_state.terminal_manager.get_buffer_manager();
    let stats = buffer_manager.get_stats().await;
    Ok(stats)
}

/// Cleanup orphaned buffers
#[tauri::command]
pub async fn cleanup_terminal_buffers(
    request: CleanupTerminalBuffersRequest,
    app_state: State<'_, AppState>,
) -> Result<(), AppError> {
    let buffer_manager = app_state.terminal_manager.get_buffer_manager();
    buffer_manager
        .cleanup_orphaned_buffers(&request.active_terminal_ids)
        .await;
    Ok(())
}

File: src/services/buffer.ts
import { api } from "./api";
import { terminalCache } from "../core/performance";

/**
 * Buffer statistics interface
 */
export interface BufferStats {
  totalTerminals: number;
  totalLines: number;
  memoryUsage: number;
}

/**
 * Get buffer as string from Rust backend
 * @param terminalId - Terminal identifier
 * @returns Promise of buffer string
 */
export async function getTerminalBuffer(terminalId: string): Promise<string> {
  return await api.call<string>("get_terminal_buffer", {
    terminalId,
  });
}

/**
 * Check if terminal has buffer in Rust backend (cached)
 * @param terminalId - Terminal identifier
 * @returns Promise of boolean
 */
export async function hasTerminalBuffer(terminalId: string): Promise<boolean> {
  return await terminalCache.hasTerminalBuffer(terminalId);
}

/**
 * Get buffer statistics from Rust backend (cached)
 * @returns Promise of buffer statistics
 */
export async function getBufferStats(): Promise<BufferStats> {
  const stats = await terminalCache.getBufferStats();
  return {
    totalTerminals: stats.totalTerminals || 0,
    totalLines: stats.totalLines || 0,
    memoryUsage: stats.memoryUsage || 0,
  };
}

/**
 * Cleanup orphaned buffers in Rust backend
 * @param activeTerminalIds - Array of active terminal IDs
 */
export async function cleanupTerminalBuffers(
  activeTerminalIds: string[],
): Promise<void> {
  return await api.call<void>("cleanup_terminal_buffers", {
    activeTerminalIds,
  });
}

/**
 * List all terminals from backend
 * @returns Promise of terminal list
 */
export async function listTerminals(): Promise<Array<{ id: string }>> {
  return await api.callRaw<Array<{ id: string }>>("list_terminals");
}
